### PR TITLE
B-167: restore master branch stability

### LINKED ISSUE #165: Cannot build from source anymore
It seems like we cannot build from the current master branch anymore (master/2fd53ed).

I've tested to build both on MacOS and Ubuntu20

## On MacOS

1. Try to build the provider
```
# go build -o /tmp/terraform-provider-opennebula
go: github.com/OpenNebula/one/src/oca/go/src/goca@v0.0.0-20210824074935-e55e3a53c4c0: missing go.sum entry; to add it:
	go mod download github.com/OpenNebula/one/src/oca/go/src/goca
go: github.com/OpenNebula/one/src/oca/go/src/goca@v0.0.0-20210824074935-e55e3a53c4c0: missing go.sum entry; to add it:
	go mod download github.com/OpenNebula/one/src/oca/go/src/goca
```

2. Install goca
```
# go mod download github.com/OpenNebula/one/src/oca/go/src/goca
```

3. Try to build the provider again
```
# go build -o /tmp/terraform-provider-opennebula
opennebula/helpers_vm.go:11:2: cannot find module providing package github.com/hashicorp/terraform/helper/schema: import lookup disabled by -mod=readonly
	(Go version in go.mod is 1.13, so vendor directory was not used.)
```

4. Try (fail) to install `terraform/helper/schema`
```
# go get -v github.com/hashicorp/terraform/helper/schema
go get: module github.com/hashicorp/terraform@upgrade found (v1.0.8), but does not contain package github.com/hashicorp/terraform/helper/schema
```

## On Ubuntu20

1. Try to build the provider
```
# root@server1:/tmp/git/terraform-provider-opennebula# go build -o terraform-provider-opennebula
go: downloading github.com/hashicorp/terraform-plugin-sdk v1.7.0
go: downloading github.com/OpenNebula/one/src/oca/go/src/goca v0.0.0-20210824074935-e55e3a53c4c0
<... output truncated ...>
go: finding github.com/hashicorp/terraform-config-inspect v0.0.0-20191115094559-17f92b0546e8
go: finding github.com/hashicorp/hcl v0.0.0-20170504190234-a4b07c25de5f
build github.com/OpenNebula/terraform-provider-opennebula: cannot load github.com/hashicorp/terraform/helper/schema: module github.com/hashicorp/terraform@latest found (v1.0.8), but does not contain package github.com/hashicorp/terraform/helper/schema
```

2. Try (fail) to install `terraform/helper/schema`
```
# root@server1:/tmp/git/terraform-provider-opennebula# go get -v github.com/hashicorp/terraform/helper/schema
go get github.com/hashicorp/terraform/helper/schema: module github.com/hashicorp/terraform@upgrade found (v1.0.8), but does not contain package github.com/hashicorp/terraform/helper/schema
```

Has there been some changes in the way to build the provider ?

### LINKED ISSUE #167: master branch unstable
After merging a lot of PRs with CI down and a lack of testing I introduced several problems: master branch didn't compile and acceptance test didn't pass anymore.
I fixed problems recently introduced in: virtual machine (around disk&nic management), virtual network address ranges

The goal of this PR is to restore master stability.

### PR DESCRIPTION
**Pull request by treywelsh**
this PR fix: 
- mainly VM disk and NIC updates code and related acceptance tests
- vm deletion in HOLD state
- a bit of refactoring to simplify the code with helper `emptyOrEqual`
- virtual network: simplify a bit AR reading, and fix locking code

### CODE REVIEW

File: opennebula/resource_opennebula_virtual_machine.go [hunk 1]
```diff
@@ -755,24 +769,66 @@ func matchNIC(NICConfig map[string]interface{}, NIC shared.NIC) bool {
...
+				return false
+			}
+			if int(sgInt) != sgConfig[i].(int) {
+				return false
+			}
+		}
+
+	}
+
+	return (NICConfig["ip"] == nil || len(NICConfig["ip"].(string)) == 0 || ip == NICConfig["ip"].(string)) &&
```
Comment (jaypif) [inline opennebula/resource_opennebula_virtual_machine.go]: Maybe create a function for readability

File: opennebula/resource_opennebula_virtual_network.go [hunk 2]
```diff
@@ -717,11 +717,11 @@ func matchARs(ARConfig map[string]interface{}, AR vn.AR) bool {
 
 	return AR.Type == ARConfig["ar_type"].(string) &&
 		AR.Size == ARConfig["size"].(int) &&
-		len(ARConfig["ip4"].(string)) > 0 || AR.IP == ARConfig["ip4"].(string) &&
-		len(ARConfig["ip6"].(string)) > 0 || AR.IP6 == ARConfig["ip6"].(string) &&
-		len(ARConfig["mac"].(string)) > 0 || AR.MAC == ARConfig["mac"].(string) &&
-		len(ARConfig["global_prefix"].(string)) > 0 || AR.GlobalPrefix == ARConfig["global_prefix"].(string) &&
-		len(ARConfig["ula_prefix"].(string)) > 0 || AR.ULAPrefix == ARConfig["ula_prefix"].(string)
+		(ARConfig["ip4"] == nil || len(ARConfig["ip4"].(string)) == 0 || AR.IP == ARConfig["ip4"].(string)) &&
```
Comment (jaypif) [inline opennebula/resource_opennebula_virtual_network.go]: here too, a function to ease reading

**Comment by treywelsh** [comment]
I applied review comments and pushed

**Comment by Th0masL** [comment]
Hi ! I can confirm that I can build from source with this PR, and also it passes the `terraform plan` step.

**Comment by treywelsh** [comment]
CI fail for ONE 6.2 should be fixed by #178

**Review by jaypif — APPROVED** [review]

**Pull request merged**